src/tools/pkcs11-tool.c

```diff
@@ -4895,26 +4910,37 @@ derive_ec_key(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE key, CK_MECHANISM_TYPE
 #endif
 
 	if (!pkey)
-		util_fatal("Cannot read EC key from %s", opt_input);
+		util_fatal("Cannot read peer EC key from %s", opt_input);
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+	key_id = EVP_PKEY_get_id(pkey);
+#else
+	key_id = EVP_PKEY_id(pkey);
+#endif
```

Comment:
can we have this abstraction in compat-ossl header file?

Reply:
Could be eliminated. OpenSC 1.1.1 and others define function `EVP_PKEY_id`
 
OpenSC 3.0,  3.1, 3.2 and 3.3 and above define function  `EVP_PKEY_get_id` but have this define: `#define EVP_PKEY_id EVP_PKEY_get_id`  

I will go ahead and use the EVP_PKEY_id